### ISSUE TITLE
Audio ducking occurs on main sound device even if NVDA Speech is set to another output

### ISSUE THREAD

**Issue by tgeczy** [nvaccess/nvda#10441]
### Steps to reproduce:
0. Ensure audio ducking is set to "when outputting speech and sounds" and a ducking-compatible synthesizer is used.
1. Start playing music or audio through another sound output device, such as USB headphones or Bluetooth. This should be chosen as the main default sound-card in Windows' sound settings.
2. Use the synthesizer dialog to change the output device to another sound card, such as your speakers.
3. Notice behavior of how audio ducks on the main sound source.

### Actual behavior:
NVDA ducks audio on main sound device only

### Expected behavior:
NVDA should duck the audio on the device currently selected for Speech output.
### Workarounds:
If the program allows for changing sound output, it's best to have it go through the secondary source device independently, leaving Windows default on the audio device speech is desired out of. NVDA does not duck audio on other sound devices beyond the primary one.

#### Other information about your system:
Used Spotify, which does not allow for the workaround to be accomplished. It must be mentioned that this is not effective for all software, especially if it does not have a built-in way to change output.
#### Does the issue still occur after restarting your PC?
Yes

**Comment by JulienCochuyt** [comment]
@tgeczy wrote:
> NVDA does not duck audio on other sound devices beyond the primary one.

While I can reproduce the main issue of sound being ducked on other devices than the one NVDA is outputting to, all sound devices are equally ducked on my system.
Tests performed:
- Configure NVDA to duck other audio when it speaks and set speech output to secondary device
- Have music come out of main and secondary devices from different applications.
Result: Music is equally ducked on all devices.

NVDA relies on Windows Accessibility Features to duck sounds.
There sadly does not seem to be a way to instruct which devices to duck and which not to.
Reference: https://docs.microsoft.com/en-us/windows/win32/api/oleacc/nf-oleacc-accsetrunningutilitystate

@michaelDCurran closing as wontfix or shall we try to forward to Microsoft?

**Comment by josephsl** [comment]
Hi, another possibility: we might not be using changes in recent Windows 10 releases. Thanks.

**Comment by tgeczy** [comment]
I hate to bring up competition, but Narrator appears to do this correctly. If I have switched the output in Settings "select where you hear Narrator's voice" to another device, it stops to duck my music from the main one.
So the question is: Are they doing something different than NVDA or accessing an API NVDA cannot?

**Comment by josephsl** [comment]
Hi, an important follow-up: is this with Windows 10 or with 8.x too? If the former, chances are that there might be an additional API or a change NVDA should be told to use in newer Windows 10 feature updates, hence I’m leaning more towards API. Thanks.

**Comment by dingpengyu** [comment]
#9416

**Comment by JulienCochuyt** [comment]
> I hate to bring up another product, but Narrator appears to do this correctly. If I have switched the output in Settings "select where you hear Narrator's voice" to another device, it stops to duck my music from the main one.
> So the question is: Are they doing something different than NVDA or accessing an API NVDA cannot?

Comparison is sometimes a good thing. Here, it shows us there probably is something we could do unless they are using a private API.

**Comment by Adriani90** [comment]
Cc: @jcsteh

**Comment by Adriani90** [comment]
cc: @ahicks92 this might be fixable easier now with WASAPI since it provides also an exclusive mode right?

**Comment by jcsteh** [comment]
This is not related to exclusive mode, nor will WASAPI fix it.

We enable ducking using AccSetRunningUtilityState, as does Narrator. However, we use a different flag than Narrator does. There are a few problems with Narrator's ducking:

1. It fades the audio up and down very fast, which can be very jarring.
2. It doesn't delay speech when it fades down, so you can miss the start of the utterance.
3. You can't do "always duck" behaviour.

So, rather than relying on Windows to determine when to duck, we manage that ourselves. Unfortunately, Windows doesn't seem to give us control over which device it ducks when we do it this way. In contrast, I guess the ducking used by Narrator knows which device to use because it figures out when to duck based on the audio stream that is playing.

I guess we could provide an option to use Narrator's ducking for users for whom this matters. It would have all of the downsides noted above, though.

**Comment by ahicks92** [comment]
Yeah, nothing to do with Wasapi.  There are maybe things that could be done, but they will almost certainly just make everything worse.

**Comment by Adriani90** [comment]
Given the lack of Windows native flexibility in this regard and that the Narator's way of ducking doesn't seem to have enough community support, I am closing as won't fix. If technology changes and there are updates in this regard, we can open a new request / discussion.

**Comment by CyrilleB79** [comment]
For this specific issue, I'd ask:
* Could the imperfect solution of Narrator's ducking be implemented only for the case where multiple audio outputs are used? Better than nothing?
* Could NV Access have a communication with MS if an API improvement is desirable?

I'd prefer NV Access to decide if closing or not such type of issues. Cc @seanbudd, @gerald-hartig, @SaschaCowley 

Reopening
